service/rpc/servicer.go

```diff
@@ -419,50 +424,228 @@ func (s *Servicer) BlockTransaction(ctx context.Context, request *types.BlockTra
+		}
+		v, r, s := t.RawSignatureValues()
+		signature := airgap.ValuesToSignature(t.ChainId(), v, r, s)
+
+		tx = airgap.Transaction{
+			TxMetadata: txMetadata,
+			Signature: signature,
+		}
+	}
+
+	ops := []*types.Operation{
+		{
+			Type: "transfer",
+			OperationIdentifier: &types.OperationIdentifier{
+				Index: 0,
+			},
+			Account: &types.AccountIdentifier{
+				Address: tx.From.Hex(),
+			},
+			Amount: &types.Amount{
+				Value:    new(big.Int).Neg(tx.Value).String(),
+				Currency: CeloGold,
+			},
+		},
+		{
+			Type: "transfer",
+			OperationIdentifier: &types.OperationIdentifier{
+				Index: 1,
+			},
+			RelatedOperations: []*types.OperationIdentifier{
+				{
+					Index: 0,
+				},
+			},
+			Account: &types.AccountIdentifier{
+				Address: tx.To.Hex(),
+			},
+			Amount: &types.Amount{
+				Value:    tx.Value.String(),
+				Currency: CeloGold,
+			},
+		},
+	}
+
+	var resp *types.ConstructionParseResponse
+	resp = &types.ConstructionParseResponse{
+		Operations:               ops,
+	}
+	if req.Signed {
+		resp.AccountIdentifierSigners = []*types.AccountIdentifier{
+			{
+				Address: tx.From.Hex(),
+			},
+		}
+	}
+	
+	return resp, nil
 }
 
 func (s *Servicer) ConstructionPayloads(
-	context.Context,
-	*types.ConstructionPayloadsRequest,
+	ctx context.Context,
+	req *types.ConstructionPayloadsRequest,
 ) (*types.ConstructionPayloadsResponse, *types.Error) {
-	return nil, LogErrUnimplemented("ConstructionPayloads")
+	bz, err := json.Marshal(req.Metadata)
+	if err != nil {
+		return nil, ErrInternal
+	}
+	var metadata airgap.TxMetadata
+	err = metadata.UnmarshalJSON(bz)
+	if err != nil {
+		return nil, ErrInternal
+	}
+
+	value := req.Operations[0].Amount.Value
+	valueInt, ok := new(big.Int).SetString(value, 10)
+	if !ok {
+		return nil, nil
+	}
+	valueInt.Abs(valueInt)
+	metadata.Value = valueInt
+
+	tx := airgap.Transaction{
+		TxMetadata: &metadata,
+		Signature: []byte{},
+	}
+	
+	gethTx, _ := tx.AsGethTransaction()
+	signer := ethTypes.NewEIP155Signer(tx.ChainId)
+	
+	// Construct SigningPayload
+	payload := &types.SigningPayload{
+		AccountIdentifier: &types.AccountIdentifier{
+			Address: tx.From.Hex(),
+		},
+		Bytes:             signer.Hash(gethTx).Bytes(),
+		SignatureType:     types.EcdsaRecovery,
+	}
+
+	unsignedTxJSON, err := json.Marshal(tx)
+	if err != nil {
+		return nil, ErrInternal
+	}
+
+	return &types.ConstructionPayloadsResponse{
+		UnsignedTransaction: string(unsignedTxJSON),
+		Payloads:            []*types.SigningPayload{payload},
+	}, nil
 }
 
 func (s *Servicer) ConstructionPreprocess(
-	context.Context,
-	*types.ConstructionPreprocessRequest,
+	ctx context.Context,
+	req *types.ConstructionPreprocessRequest,
 ) (*types.ConstructionPreprocessResponse, *types.Error) {
-	return nil, LogErrUnimplemented("ConstructionPreprocess")
+	options := make(map[string]interface{})
+	options["From"] = req.Operations[0].Account.Address
+	options["To"] = req.Operations[1].Account.Address
```

Comment:
(? checking my understanding): I'm not totally sure if we need to do anything for `preprocess`...do we actually need to make a call to `/construction/metadata` at all? according to the docs we don't need to return any options if we don't need to make a call to `/construction/metadata` and I don't understand why we would need to get From/To addresses online?

Reply:
Why would we not make a fetch to `/metadata`? These values are used in the `ConstructionMetadata` endpoint

Reply:
also not blocking merge for this proof of concept, but i think my understanding of this endpoint within the workflow is shaky. My question was more geared towards how the docs describe this endpoint's function and i think i confused myself a bit. Again not blocking, but we may also need to add handling for `Operations.amount.value` but that can come later (and prob goes nicely with the currency changes described above).